src/spdx_tools/spdx3/bump_from_spdx2/relationship.py

```diff
@@ -1,74 +1,277 @@
+    generated_relationships: Dict[Tuple[str, str], List[Relationship]] = {}
+    for counter, spdx2_relationship in enumerate(spdx2_relationships):
+        relationship = bump_relationship(spdx2_relationship, creation_information, document_namespace, counter)
+        if relationship:
+            generated_relationships.setdefault(
+                (relationship.from_element, relationship.relationship_type.name), []
+            ).append(relationship)
 
-    spdx_id = "#".join([document_namespace, f"SPDXRef-Relationship-{counter}"])
+    for key, relationships in generated_relationships.items():
+        if len(relationships) > 1:
+            merge_relationships_and_add_to_payload(payload, relationships)
+        else:
+            payload.add_element(relationships[0])
 
-    if isinstance(
-        spdx2_relationship.related_spdx_element_id, SpdxNoAssertion
-    ):  # how to translate none/ no assertion to element?
-        completeness = RelationshipCompleteness.UNKNOWN
-    elif isinstance(spdx2_relationship.related_spdx_element_id, SpdxNone):
-        completeness = RelationshipCompleteness.KNOWN
-    else:
-        completeness = None
 
-    if swap_direction:
-        from_element = spdx2_relationship.related_spdx_element_id
-        to = [spdx2_relationship.spdx_element_id]
-    else:
+def bump_relationship(
+    spdx2_relationship: Spdx2_Relationship,
+    creation_information: CreationInformation,
+    document_namespace: str,
+    counter: int,
+) -> Optional[Union[Relationship, SoftwareDependencyRelationship]]:
+    swap_direction = False
+    completeness, to = determine_completeness(spdx2_relationship.related_spdx_element_id)
+    spdx_id = "#".join([document_namespace, f"SPDXRef-Relationship-{counter}"])
+    parameters_for_bump = relationship_mapping[spdx2_relationship.relationship_type]
+    if parameters_for_bump is None:
+        print_missing_conversion(spdx2_relationship.relationship_type.name, 0)
+        return
+    base_relationship = parameters_for_bump[0]
+    relationship_type = parameters_for_bump[1]
+    if not base_relationship:
+        scope = parameters_for_bump[2]
+        software_linkage = parameters_for_bump[3]
+        conditionality = parameters_for_bump[4]
         from_element = spdx2_relationship.spdx_element_id
-        to = [spdx2_relationship.related_spdx_element_id]
 
-    payload.add_element(
-        Relationship(
+        return SoftwareDependencyRelationship(
             spdx_id,
             creation_information,
             from_element,
             to,
             relationship_type,
             comment=spdx2_relationship.comment,
             completeness=completeness,
+            scope=scope,
+            software_linkage=software_linkage,
+            conditionality=conditionality,
         )
+
+    if base_relationship and (len(parameters_for_bump) == 3):
+        swap_direction = parameters_for_bump[2]
```

Comment:
I'm confused by the parameters; in line 213 you have `scope = parameters_for_bump[2]` ...?

Reply:
This is some kind of trade-off between keeping the `relationship_mapping` as simple as possible and delaing with the different cases. So, all spdx2_relationships that can be swaooed will result in a "classic" relationship and all spdx2_relationships that result in a SoftwareDependencyRelationship need additional parameter for that. So the mapping is either spdx2_relationship -> (base_relationsihip, relationship_type, swap_direction) or spdx2_relationship -> (no_base_relationship, relationship_type, scope, software_linkage, conditionality). But we can also map each spdx2_relationship to a quintuple..